crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -0,0 +1,957 @@
+    pub fn prev_batch(&self) -> Option<String> {
+        self.prev_batch.lock_ref().clone()
+    }
+
+    /// `AliveTimeline` of this room
+    pub fn timeline(&self) -> AliveRoomTimeline {
+        self.timeline.clone()
+    }
+
+    /// This rooms name as calculated by the server, if any
+    pub fn name(&self) -> Option<String> {
+        self.inner.name.clone()
+    }
+
+    fn update(&mut self, room_data: &v4::SlidingSyncRoom) {
+        let v4::SlidingSyncRoom {
+            name,
+            initial,
+            is_dm,
+            invite_state,
+            unread_notifications,
+            required_state,
+            prev_batch,
+            timeline,
+            ..
+        } = room_data;
+
+        self.inner.unread_notifications = unread_notifications.clone();
+
+        if name.is_some() {
+            self.inner.name = name.clone();
+        }
+        if initial.is_some() {
+            self.inner.initial = *initial;
+        }
+        if is_dm.is_some() {
+            self.inner.is_dm = *is_dm;
+        }
+        if !invite_state.is_empty() {
+            self.inner.invite_state = invite_state.clone();
+        }
+        if !required_state.is_empty() {
+            self.inner.required_state = required_state.clone();
+        }
+
+        if let Some(batch) = prev_batch {
+            self.prev_batch.lock_mut().replace(batch.clone());
+        }
+
+        if !timeline.is_empty() {
+            let mut ref_timeline = self.timeline.lock_mut();
+            for e in timeline {
+                ref_timeline.push_cloned(e.clone());
+            }
+        }
+    }
+}
+
+impl std::ops::Deref for SlidingSyncRoom {
+    type Target = v4::SlidingSyncRoom;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+type ViewState = futures_signals::signal::Mutable<SlidingSyncState>;
+type SyncMode = futures_signals::signal::Mutable<SlidingSyncMode>;
+type PosState = futures_signals::signal::Mutable<Option<String>>;
+type RangeState = futures_signals::signal::Mutable<Vec<(UInt, UInt)>>;
+type RoomsCount = futures_signals::signal::Mutable<Option<u32>>;
+type RoomsList = Arc<futures_signals::signal_vec::MutableVec<RoomListEntry>>;
+type RoomsMap = Arc<futures_signals::signal_map::MutableBTreeMap<OwnedRoomId, SlidingSyncRoom>>;
+type RoomsSubscriptions =
+    Arc<futures_signals::signal_map::MutableBTreeMap<OwnedRoomId, v4::RoomSubscription>>;
+type RoomUnsubscribe = Arc<futures_signals::signal_vec::MutableVec<OwnedRoomId>>;
+type ViewsList = Arc<futures_signals::signal_vec::MutableVec<SlidingSyncView>>;
+
+use derive_builder::Builder;
+
+/// The Summary of a new SlidingSync Update received
+#[derive(Debug, Clone)]
+pub struct UpdateSummary {
+    /// The views (according to their name), which have seen an update
+    pub views: Vec<String>,
+    /// The Rooms that have seen updates
+    pub rooms: Vec<OwnedRoomId>,
+}
+
+/// The sliding sync instance
+#[derive(Clone, Debug, Builder)]
```

Comment:
With all except one field being marked `builder(private)`, wouldn't it be easier to just write out the builder by hand (if needed at all..) and remove all of these extra attributes?

Reply:
`private` does not mean, there won't be any builder params, just that they are not `pub`. They are used internally, see below. Writing all that is def. more work...

```rust startline=289
    pub fn new_builder_copy(&self) -> SlidingSyncBuilder {
        let builder = SlidingSyncBuilder::default()
            .client(self.client.clone())
            .views(Arc::new(futures_signals::signal_vec::MutableVec::new_with_values(
                self.views
                    .lock_ref()
                    .to_vec()
                    .iter()
                    .map(|v| {
                        v.new_builder().build().expect("builder worked before, builder works now")
                    })
                    .collect(),
            )))
            .subscriptions(Arc::new(futures_signals::signal_map::MutableBTreeMap::with_values(
                self.subscriptions.lock_ref().to_owned(),
            )));
          //...
```

Reply:
I see..